apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -576,28 +589,30 @@ def _create_nat_epg_for_ext_net(self, l3out_name, ext_epg_name,
                 # make external EPG use NAT contract
                 self.apic_manager.ensure_external_epg_consumed_contract(
                     l3out_name, nat_contract, external_epg=ext_epg_name,
-                    transaction=trs)
+                    owner=tenant_name, transaction=trs)
                 self.apic_manager.ensure_external_epg_provided_contract(
                     l3out_name, nat_contract, external_epg=ext_epg_name,
-                    transaction=trs)
+                    owner=tenant_name, transaction=trs)
                 # make L3-out use NAT-vrf
                 self.apic_manager.ensure_external_routed_network_created(
-                    l3out_name, context=nat_vrf_name,
+                    l3out_name, context=nat_vrf_name, owner=tenant_name,
                     transaction=trs)
 
                 # create shadow L3-out and shadow external-epg
```

Comment:
I don't think that changes what I mentioned before except for one piece of ACI behavior that needs to be validated - whether ext EPG in tenantA can refer to the NAT EPG of tenantB where tenantB is not the common tenant. Tom thinks this will work in the short term because there are no validations; however ACI may become stricter in the future. In which case, I don't see any other way than putting the L3Out for shared Ext Network in common tenant. Another option is that we keep L3Out in the specific tenant, but always create the NAT EPG/BD/VRF in the common tenant.